<a href="https://colab.research.google.com/github/LeonimerMelo/Reinforcement-Learning/blob/Policy-Gradient/REINFORCE_Policy_Gradient_Algorithm_(RL)_v2.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

#REINFORCE - Policy Gradient Algorithm

O algoritmo REINFORCE, desenvolvido por Ronald J. Williams em 1992 em seu artigo "*Simple Statistical Gradient-Following Algorithms for Connectionist Reinforcement Learning*" [[1]](https://link.springer.com/article/10.1007/BF00992696), aprende uma política parametrizada que produz probabilidades de ações a partir de estados. Os agentes usam essa política diretamente para agir em um ambiente.

A ideia central é que, durante o aprendizado, **ações que resultaram em bons resultados** devem **se tornar mais prováveis** — essas ações são positivamente reforçadas. Por outro lado, ações que resultaram em maus resultados devem se tornar menos prováveis. Se o aprendizado for bem-sucedido, ao longo de muitas iterações, as probabilidades das ações produzidas pela política se ajustam para uma distribuição que resulta em um bom desempenho no ambiente. **As probabilidades das ações são alteradas seguindo o gradiente da política**, portanto, o REINFORCE é conhecido como `algoritmo de gradiente de política`.

O algoritmo necessita de três componentes:  
1. Uma política parametrizada  
2. Um objetivo a ser maximizado  
3. Um método para atualizar os parâmetros da política  

O núcleo do algoritmo REINFORCE é o gradiente da política. O gradiente da política fornece uma maneira de estimar o gradiente do objetivo em relação aos parâmetros da política. Este é um passo crucial, pois o gradiente da política é usado para modificar os parâmetros da política de forma a maximizar o objetivo.

##Introdução
O algoritmo **REINFORCE** é uma abordagem clássica de aprendizado por reforço baseada em gradiente de política, introduzida por Ronald J. Williams em 1992. Ele é projetado para otimizar diretamente uma política parametrizada $\pi_\theta(a|s)$, onde $a$ representa uma ação, $s$ um estado, e $\theta$ são os parâmetros ajustáveis da política.

### Visão Geral do Algoritmo

1. **Política Parametrizada:**
   - O REINFORCE aprende uma política estocástica, ou seja, em vez de selecionar ações determinísticas, ele atribui uma probabilidade a cada ação em um estado específico.
   - A política é representada como $\pi_\theta(a|s)$, que é uma função parametrizada por $\theta$, frequentemente modelada por redes neurais.

2. **Objetivo:**
   - Maximizar o retorno esperado $J(\theta)$, definido como:
$$
J(\theta) = \mathbb{E}_{\pi_\theta}\left[\sum_{t=0}^\infty \gamma^t R_t\right]
$$
     onde:
     - $\gamma \in [0, 1]$ é o fator de desconto,
     - $R_t$ é a recompensa obtida no tempo $t$,
     - A expectativa é tomada sobre a distribuição de trajetórias geradas pela política $\pi_\theta$.

3. **Gradiente da Política:**
   - O REINFORCE utiliza o **teorema do gradiente da política** para derivar uma expressão para o gradiente de $J(\theta)$:
$$
\nabla_\theta J(\theta) = \mathbb{E}_{\pi_\theta}\left[\nabla_\theta \log \pi_\theta(a|s) G_t\right],
$$
onde $G_t = \sum_{k=t}^\infty \gamma^{k-t} R_k$ é o retorno acumulado a partir do tempo $t$.

4. **Atualização dos Parâmetros:**
   - Os parâmetros da política são atualizados por **gradiente ascendente**, pois o objetivo é maximizar $J(\theta)$:
$$
\theta \leftarrow \theta + \alpha \nabla_\theta \log \pi_\theta(a|s) G_t
$$
onde $\alpha$ é a taxa de aprendizado.

### Intuição do Funcionamento

- A ideia principal é reforçar ações que levam a altos retornos $G_t$, aumentando suas probabilidades. Isso é feito ajustando os parâmetros $\theta$ na direção do gradiente positivo, ponderado pelas recompensas acumuladas.
- **Log-Probabilidade:** A inclusão do termo $\log \pi_\theta(a|s)$ conecta diretamente a probabilidade de escolher uma ação aos retornos observados, permitindo que o aprendizado seja direcionado.

### Benefícios

1. **Simplicidade:**
   - O REINFORCE é conceitualmente simples e direto, tornando-o um ponto de partida comum no aprendizado por reforço baseado em gradientes.
2. **Exploração Automática:**
   - Como a política é estocástica, o REINFORCE explora naturalmente diferentes ações, sem necessidade de estratégias adicionais de exploração.

### Desafios

1. **Alta Variância no Gradiente:**
   - O retorno $G_t$ pode ser ruidoso, especialmente em ambientes com recompensas esparsas, levando a uma alta variância na estimativa do gradiente.
   - Isso pode causar instabilidade e tornar o treinamento mais lento.

2. **Recompensas Não Escaladas:**
   - Recompensas de diferentes magnitudes podem afetar a estabilidade do aprendizado. Normalizar recompensas ou usar métodos de base pode mitigar esse problema.

3. **Dependência de Fator de Desconto ($\gamma$):**
   - A escolha de $\gamma$ impacta fortemente o aprendizado, pois controla o equilíbrio entre recompensas imediatas e futuras.

### Extensões e Melhorias

Vários algoritmos foram desenvolvidos para superar as limitações do REINFORCE:

1. **Baseline Subtraction:**
   - Reduz a variância subtraindo uma baseline $b(s)$ do retorno $G_t$, onde $b(s)$ é tipicamente uma estimativa do valor do estado:
$$
\nabla_\theta J(\theta) = \mathbb{E}_{\pi_\theta}\left[\nabla_\theta \log \pi_\theta(a|s) (G_t - b(s))\right]
$$

2. **Algoritmos Advantage Actor-Critic (A2C/A3C):**
   - Introduzem a vantagem $A(s, a) = G_t - V(s)$, onde $V(s)$ é o valor esperado do estado $s$. Isso melhora a eficiência do aprendizado.

3. **Trust Region Policy Optimization (TRPO) e Proximal Policy Optimization (PPO):**
   - Melhoram a estabilidade ao restringir as atualizações da política dentro de limites controlados.

### Aplicações

O REINFORCE é usado em tarefas de controle e otimização, incluindo:
- Controle robótico.
- Jogos (como aprendizado em ambientes da OpenAI Gym).
- Sistemas de recomendação com aprendizado de políticas.

Apesar de suas limitações, o REINFORCE desempenha um papel fundamental na história e desenvolvimento do aprendizado por reforço baseado em gradientes.

**Tradução para o português:**

### 2.1 Política  
Uma política $\pi$ é uma função que mapeia estados para probabilidades de ações, utilizada para amostrar uma ação $a \sim \pi(s)$. No REINFORCE, um agente aprende uma política e a utiliza para agir em um ambiente.

Uma boa política é aquela que maximiza as recompensas acumuladas. A ideia central do algoritmo é aprender uma boa política, o que implica realizar aproximação de funções. Redes neurais são aproximadores de funções poderosos e flexíveis, por isso podemos representar uma política usando uma rede neural profunda composta por parâmetros aprendíveis $\theta$. Isso é frequentemente chamado de **rede de política** $\pi_\theta$. Dizemos que a política é parametrizada por $\theta$.

Cada conjunto específico de valores dos parâmetros da rede de política representa uma política particular. Para entender por que, considere $\theta_1 \neq \theta_2$. Para um dado estado $s$, redes de políticas diferentes podem gerar conjuntos diferentes de probabilidades de ações, ou seja, $\pi_{\theta_1}(s) \neq \pi_{\theta_2}(s)$. Os mapeamentos de estados para probabilidades de ações são diferentes, então dizemos que $\pi_{\theta_1}$ e $\pi_{\theta_2}$ são políticas diferentes. Uma única rede neural, portanto, é capaz de representar muitas políticas diferentes.

Formulada dessa maneira, **o processo de aprender uma boa política corresponde a buscar um bom conjunto de valores para** $\theta$. Por essa razão, é importante que a rede de política seja diferenciável. O mecanismo pelo qual a política é aprimorada é por meio da **ascensão do gradiente** no espaço dos parâmetros.

##Função Objetivo  
Agora definimos o objetivo maximizado por um agente no algoritmo REINFORCE. Um objetivo pode ser entendido como a meta de um agente, como vencer um jogo ou obter a maior pontuação possível. Primeiramente, introduzimos o conceito de retorno, que é calculado usando uma trajetória. Em seguida, usamos isso para formular o objetivo.

Um agente atuando em um ambiente gera uma **trajetória**, que contém uma sequência de recompensas juntamente com os estados e ações. Uma trajetória é denotada por:  
$$
\tau = s_0, a_0, r_0, \ldots, s_T, a_T, r_T
$$

O **retorno** de uma trajetória $R_t(\tau)$ é definido como a soma de recompensas descontadas desde o instante de tempo $t$ até o final da trajetória, conforme mostrado na Equação:  
$$
R_t(\tau) = \sum_{t' = t}^T \gamma^{t' - t} r_{t'}
$$

Observe que a soma começa no instante de tempo $t$, mas a potência do fator de desconto $\gamma$ começa em 0. Por isso, é necessário ajustar a potência com base no instante inicial $t$, usando $t' - t$.

Quando $t = 0$, o retorno é simplesmente o retorno da trajetória completa. Para simplificar, isso também é escrito como $R_0(\tau) = R(\tau)$.  

O **objetivo** é o retorno esperado sobre todas as trajetórias completas geradas por um agente. Isso é definido na Equação:  
$$
J(\pi_\theta) = \mathbb{E}_{\tau \sim \pi_\theta}[R(\tau)] = \mathbb{E}_{\tau \sim \pi_\theta}\left[\sum_{t=0}^T \gamma^t r_t \right]
$$
A Equação acima afirma que a expectativa é calculada sobre muitas trajetórias amostradas a partir de uma política, ou seja, $\tau \sim \pi_\theta$. Essa expectativa se aproxima do valor verdadeiro à medida que mais amostras são coletadas, e está vinculada à política específica $\pi_\theta$ utilizada.

### Explicação Intuitiva  

- **Trajetória ($\tau$)**: Representa o caminho que o agente percorre no ambiente, incluindo estados ($s$), ações ($a$) e recompensas ($r$).  
- **Retorno ($R_t(\tau)$)**: Mede o valor total das recompensas futuras (descontadas) que o agente espera a partir do instante $t$.  
- **Função Objetivo ($J(\pi_\theta)$)**: O objetivo do agente é maximizar o retorno esperado, ou seja, aprender uma política que gere trajetórias com altos valores acumulados de recompensa.

## O Gradiente da Política  

Agora definimos a política $\pi_\theta$ e o objetivo $J(\pi_\theta)$, que são dois componentes cruciais para a derivação do algoritmo de gradiente de política. **A política fornece uma maneira para o agente agir, e o objetivo estabelece um alvo a ser maximizado**.

O componente final do algoritmo é o **gradiente da política**. Formalmente, dizemos que um algoritmo de gradiente de política resolve o seguinte problema:  
$$
\max_{\theta} J(\pi_\theta) = \mathbb{E}_{\tau \sim \pi_\theta}[R(\tau)]
$$

Para maximizar o objetivo, realizamos a **subida por gradiente** nos parâmetros da política $\theta$. Lembre-se, do cálculo diferencial, que o gradiente aponta na direção de maior subida. Para melhorar o objetivo, calculamos o gradiente e o utilizamos para atualizar os parâmetros, como mostrado na Equação:  
$$
\theta \leftarrow \theta + \alpha \nabla_\theta J(\pi_\theta)
$$  
onde:
- $\alpha$ é a taxa de aprendizado,
- $\nabla_\theta J(\pi_\theta)$ é o gradiente do objetivo em relação aos parâmetros da política.

Essa atualização ajusta os parâmetros $\theta$ na direção que maximiza o retorno esperado $J(\pi_\theta)$.

O termo $\nabla_\theta J(\pi_\theta)$ é conhecido como o **gradiente da política**. Ele é definido na Equação:  
$$
\nabla_\theta J(\pi_\theta) = \mathbb{E}_{\tau \sim \pi_\theta}\left[\sum_{t=0}^T R_t(\tau) \nabla_\theta \log \pi_\theta(a_t | s_t)\right]
$$

<!--\tag{2.5} para colocar número nas equações  -->

O termo $\pi_\theta(a_t | s_t)$ é a probabilidade da ação tomada pelo agente no passo de tempo $t$. A ação é amostrada da política, $a_t \sim \pi_\theta(s_t)$. O lado direito da equação afirma que o gradiente do logaritmo da probabilidade da ação em relação a $\theta$ é multiplicado pelo retorno $R_t(\tau)$.

A Equação afirma que o gradiente do objetivo é equivalente à soma esperada dos gradientes dos logaritmos das probabilidades das ações $a_t$, multiplicada pelos respectivos retornos $R_t(\tau)$. A derivação completa é mostrada mais a frente.

O **gradiente da política** é o mecanismo pelo qual as probabilidades das ações produzidas pela política são alteradas. Se o retorno $R_t(\tau) > 0$, então a probabilidade da ação $\pi_\theta(a_t | s_t)$ é aumentada; inversamente, se o retorno $R_t(\tau) < 0$, então a probabilidade da ação $\pi_\theta(a_t | s_t)$ é diminuída. Ao longo de muitas atualizações
da Equação, a política aprenderá a produzir ações que resultam em altos valores de $R_t(\tau)$.

A Equação é a base de todos os métodos de gradiente de política. O REINFORCE foi o primeiro algoritmo a utilizá-la em sua forma mais simples. Algoritmos mais novos implementam com base nisto, modificando a função para melhorar o desempenho.

##Amostragem Monte Carlo
O algoritmo **REINFORCE** estima numericamente o gradiente da política usando amostragem Monte Carlo.

A amostragem Monte Carlo refere-se a qualquer método que utiliza amostragem aleatória para gerar dados usados para aproximar uma função. Em essência, é apenas "aproximação com amostragem aleatória". É uma técnica que se popularizou graças a Stanislaw Ulam, um matemático que trabalhou no laboratório de pesquisa de Los Alamos na década de 1940.

Para entender como a amostragem Monte Carlo funciona, vamos passar por um exemplo de como ela pode ser usada para estimar o valor de $\pi$ (a constante matemática)—a razão entre a circunferência e o diâmetro de um círculo. Uma abordagem de Monte Carlo para resolver esse problema é pegar um círculo de raio $r = 1$ centrado na origem e inscrevê-lo em um quadrado. As áreas do círculo e do quadrado são $\pi r^2$ e $(2r)^2$, respectivamente. Assim, a razão entre essas áreas é simplesmente:
$$
\frac{\text{área do círculo}}{\text{área do quadrado}} = \frac{\pi r^2}{(2r)^2} = \frac{\pi}{4}
$$

Numericamente, o quadrado tem uma área de 4, mas como ainda não sabemos o valor de $\pi$, a área do círculo é desconhecida. Vamos considerar um quadrante. Para obter uma estimativa de $\pi$, amostramos muitos pontos dentro do quadrado usando uma distribuição aleatória uniforme. Um ponto $(p_x, p_y)$ que cai dentro do círculo tem uma distância menor que 1 da origem—ou seja, $(x - 0)^2 + (y - 0)^2 \leq 1$. Isso é mostrado na Figura abaixo. Usando isso, se contarmos o número de pontos dentro do círculo e depois contarmos o número total de pontos amostrados, a razão entre eles será aproximadamente igual à Equação acima. Ao amostrar iterativamente mais pontos e atualizar a razão, nossa estimativa se aproximará do valor preciso. Multiplicando essa razão por 4, obtemos o valor estimado de $\pi \approx 3,14159$.

<center>
<img src='https://drive.google.com/uc?id=1gtT1z2QWps_SCsyYhsVp7NkE_L6lcOXj' width=300>
</center>

In [ ]:
import numpy as np
t_circ = 0
t_quad = 100000
for i in range(t_quad):
  x = np.random.rand()
  y = np.random.rand()
  d = np.sqrt(x**2 + y**2)
  if d <= 1:
    t_circ += 1
pi = 4 * t_circ / t_quad
print('pi ~', pi)

pi ~ 3.14148


Agora, vamos voltar ao **Deep RL** e ver como a amostragem **Monte Carlo** pode ser usada para estimar numericamente o gradiente da política. É bastante simples. A expectativa $ \mathbb{E}_{\tau \sim \pi_\theta} $ implica que, à medida que mais trajetórias $ \tau $ são amostradas usando uma política $ \pi_\theta $ e calculadas a média, elas se aproximam do gradiente real da política $ \nabla_\theta J(\pi_\theta) $. Em vez de amostrar muitas trajetórias por política, podemos amostrar apenas uma, como mostrado na Equação:
$$
\nabla_\theta J(\pi_\theta) \approx \sum_{t=0}^T R_t(\tau) \nabla_\theta \log \pi_\theta(a_t | s_t)
$$

É assim que o **gradiente da política** é implementado—como uma estimativa Monte Carlo sobre trajetórias amostradas.

##Algoritmo **REINFORCE**

O algoritmo REINFORCE é bastante simples. Primeiro, inicializa-se a taxa de aprendizado $\alpha$ e constrói-se uma rede de política $\pi_\theta$ com pesos inicializados aleatoriamente.  

Em seguida, itera-se por múltiplos episódios da seguinte maneira: usa-se a rede de política $\pi_\theta$ para gerar uma trajetória $\tau = s_0, a_0, r_0, \dots, s_T, a_T, r_T$ para um episódio. Depois, para cada passo de tempo $t$ na trajetória, calcula-se o retorno $R_t(\tau)$. Usa-se $R_t(\tau)$ para estimar o gradiente da política. Soma-se os gradientes da política para todos os passos de tempo e então usa-se o resultado para atualizar os parâmetros da rede de política $\theta$.

**Algoritmo REINFORCE**:

- Inicialize a taxa de aprendizado $\alpha$
- Inicialize os pesos $\theta$ de uma rede de política $\pi_\theta$
- for episódio = 0, ..., MAX_EPISODE do:
  - Amostre uma trajetória $\tau = s_0, a_0, r_0, \dots, s_T, a_T, r_T$
  - set $\nabla_\theta J(\pi_\theta) = 0$
  - for $t = 0, ..., T$ do:
    - $R_t(\tau) = \sum_{t_0=t}^{T} \gamma^{t_0 - t} r_{t_0}$
    - $\nabla_\theta J(\pi_\theta) = \nabla_\theta J(\pi_\theta) + R_t(\tau) \nabla_\theta \log \pi_\theta(a_t | s_t)$
  - end for
  - $\theta = \theta + \alpha \nabla_\theta J(\pi_\theta)$
- end for

É importante que uma trajetória seja descartada após cada atualização de parâmetros — ela não pode ser reutilizada. Isso ocorre porque o **REINFORCE** é um algoritmo **on-policy**. Um algoritmo é **on-policy** se a equação de atualização dos parâmetros depende da política atual. Isso é mostrado na atualização do gradiente, pois o gradiente da política depende diretamente das probabilidades de ação $\pi_\theta(a_t | s_t)$ geradas pela política atual $\pi_\theta$, e não de alguma política passada $\pi_\theta'$.

De forma correspondente, o retorno $R_t(\tau)$, onde $\tau \sim \pi_\theta$, também deve ser gerado pela $\pi_\theta$, caso contrário, as probabilidades de ação serão ajustadas com base em retornos que a política não teria gerado.

## Melhorando o REINFORCE

Nossa formulação do algoritmo **REINFORCE** estima o gradiente da política usando amostragem **Monte Carlo** com uma única trajetória. Essa é uma estimativa não tendenciosa do gradiente da política, mas uma desvantagem dessa abordagem é que ela tem uma alta **variância**. Veremos uma **linha de base** para reduzir a variância da estimativa. Após isso, também discutiremos a **normalização das recompensas** para lidar com o problema de **escala das recompensas**.

Ao usar amostragem Monte Carlo, a estimativa do gradiente da política pode ter alta variância porque os retornos podem variar significativamente de trajetória para trajetória. Isso ocorre devido a três fatores. Primeiro, as ações possuem uma certa aleatoriedade, pois são amostradas de uma distribuição de probabilidade. Segundo, o estado inicial pode variar a cada episódio. Terceiro, a função de transição do ambiente pode ser estocástica.

Uma maneira de reduzir a variância da estimativa é modificar os retornos subtraindo uma linha de base adequada, independente da ação, como mostrado na Equação:

$$
\nabla_\theta J(\pi_\theta) \approx \sum_{t=0}^T \left( R_t(\tau) - b(s_t) \right) \nabla_\theta \log \pi_\theta(a_t | s_t)
$$
onde $b(s_t)$ é a linha de base, uma função que depende apenas do estado $s_t$ e não da ação tomada.

Uma opção para a linha de base é a função de valor $V^\pi$. Essa escolha de linha de base motiva o algoritmo Actor-Critic, uma técnica fundamental em aprendizado por reforço. Nesse contexto, a função de valor $V^\pi$ é usada como uma linha de base para reduzir a variância nas estimativas do gradiente. Essa abordagem é benéfica porque ajuda a melhorar a estabilidade e a eficiência do processo de aprendizado.

Nos métodos Actor-Critic, o "Ator" refere-se à política $ \pi $, que determina as ações do agente, enquanto o "Crítico" refere-se à função de valor $ V^\pi $, que avalia quão boa é a ação escolhida com base na política atual. Ao usar $ V^\pi $ como uma linha de base, o crítico fornece um feedback ao ator, permitindo que o agente melhore sua política ao longo do tempo, ajustando as ações que realiza com base nas recompensas recebidas.

Essa estrutura divide a tarefa em dois componentes:
1. **Ator**: Atualiza a política (o processo de tomada de decisão) com base no feedback do crítico.
2. **Crítico**: Estima a função de valor (quão boa é a ação) e ajuda a reduzir a variância das atualizações da política.

A vantagem de usar a função de valor como linha de base é que ela reduz a variância do gradiente, o que pode levar a um aprendizado mais estável e rápido em comparação com métodos puros de gradiente de política.

Uma outra alternativa é usar os retornos médios ao longo da trajetória. Seja $ b = \frac{1}{T} \sum_{t=0}^{T} R_t(\tau) $.

Observe que essa é uma linha de base constante por trajetória, que não varia com o estado $ s_t $. Ela tem o efeito de centralizar os retornos para cada trajetória em torno de 0. Para cada trajetória, em média, as melhores 50% das ações serão incentivadas, enquanto as outras serão desencorajadas.

##**REINFORCE** em Python
Exemplo simples de implementação do algoritmo **REINFORCE** em Python, que é uma abordagem de **Policy Gradient** para treinamento de agentes em ambientes de reforço. Este código utiliza rede neural básica configurado com uma biblioteca como o **PyTorch** ou **TensorFlow**.

##REINFORCE com **PyTorch**

### Requisitos:
- **PyTorch** instalado (`pip install torch`).
- Um ambiente do OpenAI Gym para testar, como `CartPole-v1` (que pode ser instalado com `pip install gym`).

### Exemplo de Código:

```python
import torch
import torch.nn as nn
import torch.optim as optim
import gymnasium as gym
import numpy as np

# Definindo a rede neural para a política (actor)
class PolicyNetwork(nn.Module):
    def __init__(self, input_size, output_size):
        super(PolicyNetwork, self).__init__()
        self.fc1 = nn.Linear(input_size, 128)
        self.fc2 = nn.Linear(128, output_size)
        self.softmax = nn.Softmax(dim=-1)

    def forward(self, x):
        x = torch.relu(self.fc1(x))
        x = self.fc2(x)
        return self.softmax(x)

# REINFORCE Algorithm
def reinforce(env, policy_network, optimizer, episodes=1000, gamma=0.99):
    for episode in range(episodes):
        state, _ = env.reset()
        done = False
        log_probs = []
        rewards = []

        while not done:
            state = torch.tensor(state, dtype=torch.float32)
            action_probs = policy_network(state)
            dist = torch.distributions.Categorical(action_probs)
            action = dist.sample()
            
            log_probs.append(dist.log_prob(action))
            state, reward, done, truncated, _ = env.step(action.item())
            rewards.append(reward)

        # Calcular o retorno descontado (G_t)
        returns = []
        R = 0
        for r in rewards[::-1]:
            R = r + gamma * R
            returns.insert(0, R)

        # Normalizando os retornos para estabilidade
        returns = torch.tensor(returns)
        returns = (returns - returns.mean()) / (returns.std() + 1e-5)

        # Atualizar a política usando o gradiente
        loss = -torch.sum(torch.stack(log_probs) * returns)  # REINFORCE Update Rule
        optimizer.zero_grad()
        loss.backward()
        optimizer.step()

        if episode % 100 == 0:
            print(f"Episode {episode}/{episodes}, Loss: {loss.item()}")

# Configuração do ambiente e do modelo
env = gym.make('CartPole-v1')
input_size = env.observation_space.shape[0]
output_size = env.action_space.n

policy_network = PolicyNetwork(input_size, output_size)
optimizer = optim.Adam(policy_network.parameters(), lr=1e-2)

# Treinando o agente com REINFORCE
reinforce(env, policy_network, optimizer, episodes=1000)
```

### Explicação:

1. **`PolicyNetwork`**: Define uma rede neural simples com uma camada oculta para a política, que retorna a distribuição de probabilidade de ações possíveis.
2. **`reinforce`**: Implementa o algoritmo REINFORCE. Em cada episódio:
   - O agente escolhe ações com base nas probabilidades da política atual.
   - Ele acumula os logaritmos das probabilidades das ações tomadas (`log_probs`).
   - Ele acumula as recompensas de cada ação para calcular o retorno descontado (`returns`).
3. **Função de Perda**: A perda é calculada como o produto das probabilidades logarítmicas das ações tomadas e os retornos descontados. A ideia é maximizar a expectativa dos retornos ao longo das ações tomadas.
4. **Treinamento**: O modelo é treinado usando o otimizador Adam e a função de perda é minimizada.

### Como funciona:

- O agente coleta uma sequência de estados, ações e recompensas.
- Ele calcula o retorno total para cada ação tomada.
- O modelo é atualizado com base na recompensa futura esperada para as ações passadas.
- O algoritmo REINFORCE usa o gradiente das políticas para ajustar a política do agente.

Esse é um exemplo básico que pode ser adaptado, conforme necessário, para ambientes mais complexos ou redes neurais mais profundas.

## Explicação mais detalhada
Este código é uma implementação do algoritmo **REINFORCE** em **PyTorch**, que utiliza um modelo baseado em gradientes de política para treinar um agente no ambiente `CartPole-v1` do **OpenAI Gym**. Vamos detalhar cada parte:

---

### **1. Importação de bibliotecas**
```python
import torch
import torch.nn as nn
import torch.optim as optim
import gymnasium as gym
import numpy as np
```
- **`torch`**: Usado para construir e treinar redes neurais, lidar com tensores e operações de gradiente.
- **`torch.nn`**: Contém módulos para definir modelos, como camadas de redes neurais.
- **`torch.optim`**: Inclui otimizadores para atualizar os pesos do modelo.
- **`gymnasium`**: Fornece um conjunto de ambientes simulados para aprendizado por reforço, como `CartPole-v1`.
- **`numpy`**: Usado para cálculos numéricos, como manipulação de listas e arrays.

---

### **2. Definição da Rede Neural**
```python
class PolicyNetwork(nn.Module):
    def __init__(self, input_size, output_size):
        super(PolicyNetwork, self).__init__()
        self.fc1 = nn.Linear(input_size, 128)  # Camada totalmente conectada (entrada -> 128)
        self.fc2 = nn.Linear(128, output_size) # Camada totalmente conectada (128 -> saída)
        self.softmax = nn.Softmax(dim=-1)     # Camada softmax para converter logits em probabilidades

    def forward(self, x):
        x = torch.relu(self.fc1(x))           # Aplicando ReLU após a primeira camada
        x = self.fc2(x)                       # Passando pela segunda camada
        return self.softmax(x)                # Retornando probabilidades de ação
```
- **`PolicyNetwork`**: Define a rede neural para o agente (também chamado de "ator"). Ela recebe o estado atual e retorna as probabilidades de selecionar cada ação.
- **Estrutura**:
  - Uma camada oculta com 128 unidades e ativação ReLU.
  - Uma camada de saída com `output_size` neurônios (um para cada ação possível).
  - A saída é passada por uma função softmax para garantir que as probabilidades somem 1.

---

### **3. Função REINFORCE**
```python
def reinforce(env, policy_network, optimizer, episodes=1000, gamma=0.99):
```
Essa função implementa o algoritmo REINFORCE, que treina o agente com base no gradiente da política.

#### **3.1 Loop principal de episódios**
```python
    for episode in range(episodes):
        state, _ = env.reset()  # Reinicia o ambiente no início do episódio
        done = False
        log_probs = []       # Lista para armazenar logaritmos das probabilidades de ações
        rewards = []         # Lista para armazenar recompensas do episódio
```

- **`env.reset()`**: Reinicia o ambiente e retorna o estado inicial.
- **`log_probs`**: Armazena os logaritmos das probabilidades das ações escolhidas.
- **`rewards`**: Armazena as recompensas recebidas em cada etapa do episódio.

---

#### **3.2 Loop interno (interações com o ambiente)**
```python
        while not done:
            state = torch.tensor(state, dtype=torch.float32)  # Converte o estado para tensor
            action_probs = policy_network(state)              # Obtém probabilidades de ação
            dist = torch.distributions.Categorical(action_probs)  # Define distribuição categórica
            action = dist.sample()                            # Amostra uma ação da distribuição
            
            log_probs.append(dist.log_prob(action))           # Armazena log(probabilidade da ação)
            state, reward, done, truncated, _ = env.step(action.item())  # Executa a ação no ambiente
            rewards.append(reward)                            # Armazena a recompensa
```
- A cada interação:
  - O estado atual é convertido para tensor.
  - A rede neural retorna as probabilidades das ações possíveis.
  - Uma ação é escolhida amostrando da distribuição categórica gerada pelas probabilidades.
  - A ação é executada no ambiente, e o próximo estado, a recompensa e o indicador `done` (se o episódio terminou) são retornados.

---

#### **3.3 Cálculo do retorno descontado**
```python
        returns = []
        R = 0
        for r in rewards[::-1]:
            R = r + gamma * R
            returns.insert(0, R)  # Insere o retorno acumulado na posição 0
```
- **`returns`**: Armazena os retornos acumulados ($ G_t $) para cada etapa.
- **`gamma`**: Fator de desconto, controla a importância das recompensas futuras.
- O retorno é calculado de trás para frente para simplificar a soma acumulada $R_t = r_t + \gamma R_{t+1}$.

---

#### **3.4 Normalização dos retornos**
```python
        returns = torch.tensor(returns)
        returns = (returns - returns.mean()) / (returns.std() + 1e-5)
```
- **Por que normalizar?**
  - Para estabilizar o treinamento, já que valores extremos podem levar a gradientes instáveis.
  - Normaliza os retornos para que tenham média 0 e desvio padrão 1.

---

#### **3.5 Atualização do modelo**
```python
        loss = -torch.sum(torch.stack(log_probs) * returns)  # Calcula a perda
        optimizer.zero_grad()                               # Zera os gradientes acumulados
        loss.backward()                                     # Computa os gradientes
        optimizer.step()                                    # Atualiza os pesos do modelo
```
- **Perda**:
  - $ \text{loss} = - \sum_t \log(\pi(a_t | s_t)) \cdot G_t $
  - Maximiza a probabilidade das ações que levaram a altos retornos.
- **Otimizador**:
  - Atualiza os pesos da rede neural com base nos gradientes calculados.

---

#### **3.6 Log de progresso**
```python
        if episode % 100 == 0:
            print(f"Episode {episode}/{episodes}, Loss: {loss.item()}")
```
- Exibe o progresso a cada 100 episódios, incluindo o valor da função de perda.

---

### **4. Configuração e Treinamento**
```python
env = gym.make('CartPole-v1')  # Cria o ambiente
input_size = env.observation_space.shape[0]  # Dimensão do estado
output_size = env.action_space.n             # Número de ações possíveis

policy_network = PolicyNetwork(input_size, output_size)  # Cria a rede de política
optimizer = optim.Adam(policy_network.parameters(), lr=1e-2)  # Define o otimizador

reinforce(env, policy_network, optimizer, episodes=1000)  # Treina o agente
```
- **Ambiente**: `CartPole-v1`, onde o objetivo é equilibrar o pêndulo em um carrinho.
- **Rede Neural**: Inicializada com a entrada e a saída correspondentes ao ambiente.
- **Treinamento**: O agente é treinado por 1000 episódios.

---

### **Resumo**
Este código treina um agente no ambiente `CartPole-v1` usando o algoritmo REINFORCE. Ele usa uma rede neural para modelar a política, coleta recompensas de interações com o ambiente e ajusta a política para maximizar os retornos acumulados.

## REINFORCE com TensorFlow:
Aqui temos a implementação do algoritmo **REINFORCE** usando **TensorFlow** em vez de **PyTorch**. O código tem uma estrutura semelhante, mas usa os recursos do TensorFlow para definir a rede neural e calcular o gradiente.

```python
import tensorflow as tf
import gymnasium as gym
import numpy as np

# Definindo a rede neural para a política (actor)
class PolicyNetwork(tf.keras.Model):
    def __init__(self, input_size, output_size):
        super(PolicyNetwork, self).__init__()
        self.dense1 = tf.keras.layers.Dense(128, activation='relu')
        self.dense2 = tf.keras.layers.Dense(output_size, activation='softmax')

    def call(self, inputs):
        x = self.dense1(inputs)
        return self.dense2(x)

# REINFORCE Algorithm
def reinforce(env, policy_network, optimizer, episodes=1000, gamma=0.99):
    for episode in range(episodes):
        state, _ = env.reset()
        done = False
        log_probs = []
        rewards = []

        while not done:
            state = tf.convert_to_tensor(state, dtype=tf.float32)  # Convertendo para tensor
            action_probs = policy_network(state[None, :])  # Adicionando uma dimensão extra
            action = np.random.choice(len(action_probs[0]), p=action_probs[0].numpy())  # Selecionando ação
            log_prob = tf.math.log(action_probs[0, action])  # Log da probabilidade da ação
            log_probs.append(log_prob)

            state, reward, done, truncated, _ = env.step(action)
            rewards.append(reward)

        # Calcular o retorno descontado (G_t)
        returns = []
        R = 0
        for r in rewards[::-1]:
            R = r + gamma * R
            returns.insert(0, R)

        # Normalizando os retornos para estabilidade
        returns = np.array(returns)
        returns = (returns - np.mean(returns)) / (np.std(returns) + 1e-5)

        # Atualizar a política usando o gradiente
        with tf.GradientTape() as tape:
            loss = -tf.reduce_sum(tf.convert_to_tensor(log_probs) * tf.convert_to_tensor(returns))  # REINFORCE Update Rule

        grads = tape.gradient(loss, policy_network.trainable_variables)
        optimizer.apply_gradients(zip(grads, policy_network.trainable_variables))

        if episode % 100 == 0:
            print(f"Episode {episode}/{episodes}, Loss: {loss.numpy()}")

# Configuração do ambiente e do modelo
env = gym.make('CartPole-v1')
input_size = env.observation_space.shape[0]
output_size = env.action_space.n

policy_network = PolicyNetwork(input_size, output_size)
optimizer = tf.keras.optimizers.Adam(learning_rate=1e-2)

# Treinando o agente com REINFORCE
reinforce(env, policy_network, optimizer, episodes=1000)
```

### Explicação das Alterações para TensorFlow:

1. **Rede Neural**:
   - Usamos a classe `tf.keras.Model` para definir a rede neural.
   - Em vez de `torch.nn.Module` (como no PyTorch), definimos a rede com `tf.keras.layers.Dense` e ativação `relu` para a camada oculta, e `softmax` na camada de saída para representar as probabilidades de ações.

2. **Função `call`**:
   - O método `call` é utilizado para definir o fluxo da rede neural. Ele é equivalente ao método `forward` em PyTorch.
   
3. **Conversão para TensorFlow Tensors**:
   - Para garantir que os estados sejam compatíveis com o TensorFlow, usamos `tf.convert_to_tensor(state, dtype=tf.float32)` para converter o estado em um tensor de tipo `float32`.
   - A função `policy_network(state[None, :])` é usada para adicionar uma dimensão extra ao vetor de estado, já que o modelo espera um batch de estados, e aqui estamos passando apenas um estado de cada vez.

4. **Seleção de Ação**:
   - A ação é escolhida com base nas probabilidades retornadas pela rede usando `np.random.choice` e as probabilidades são passadas como `p=action_probs[0].numpy()`.

5. **Cálculo do Gradiente e Atualização**:
   - O cálculo da perda e do gradiente é feito dentro de um bloco `tf.GradientTape()`. Depois, usamos `optimizer.apply_gradients` para aplicar os gradientes à rede neural.

6. **Função de Perda**:
   - A perda é calculada como a soma dos produtos dos logaritmos das probabilidades das ações com os retornos descontados (`log_probs` e `returns`), semelhante ao cálculo no PyTorch.

### Testando:

Para usar esse código, basta instalar as dependências necessárias:

- **TensorFlow**: `pip install tensorflow`
- **Gym**: `pip install gymnasium`

Isso deve permitir que você treine um agente no ambiente `CartPole-v1` usando o algoritmo REINFORCE com TensorFlow. Experimente rodar o código e veja como o agente aprende a equilibrar o pêndulo!

##Referências
[1] Dong H. **Deep Reinforcement Learning. Fundamentals, Research and Applications**. Springer. 2020.

[2] Laura Graesser, Wah Loon Keng. **Foundations of Deep Reinforcement Learning - Theory and Practice in Python**. Pearson Addison-Wesley. 2023.

[3] https://aws.amazon.com/pt/what-is/reinforcement-learning/

[4] https://www.deeplearningbook.com.br/o-que-e-aprendizagem-por-reforco/

[5] https://www.geeksforgeeks.org/what-is-reinforcement-learning/

[6] https://pytorch.org/tutorials/intermediate/reinforcement_q_learning.html

[7] https://www.datacamp.com/pt/tutorial/reinforcement-learning-python-introduction

[8] https://towardsdatascience.com/reinforcement-learning-rl-101-with-python-e1aa0d37d43b

[9] https://machinelearningmastery.com/principles-of-reinforcement-learning-an-introduction-with-python/

[10] https://medium.com/turing-talks/aprendizado-por-refor%C3%A7o-1-introdu%C3%A7%C3%A3o-7382ebb641ab

[11] https://www.datacamp.com/tutorial/reinforcement-learning-python-introduction

[12] https://gymnasium.farama.org/introduction/basic_usage/

[13] Sudharsan Ravichandiran. **Deep Reinforcement Learning with Python**. Second Edition. Packt Publishing. 2020.

[14] Morales, M. **Grokking Deep Reinforcement Learning**. Version 11. Manning Publications. 2020.

[15] https://www.learndatasci.com/tutorials/reinforcement-q-learning-scratch-python-openai-gym/

[16] https://github.com/sudharsan13296/Deep-Reinforcement-Learning-With-Python/tree/master